### PR TITLE
feat(any): Add publisher for media_file, album and artists - 389

### PR DESCRIPTION
**Pull request by raminou**
Import from scanning and display in the UI with filtering

Implementation of "publisher" very similar to "genre"

https://github.com/navidrome/navidrome/issues/389

screenshot_navidrome_publisher_list

screenshot_navidrome_publisher_filter

### CODE REVIEW

**Comment by raminou** [comment]
Resolved conflicts & squashed

File: model/mediafile.go [hunk 1]
```diff
@@ -159,6 +162,9 @@ func (mfs MediaFiles) ToAlbum() Album {
 	a.Genre = slice.MostFrequent(a.Genres).Name
 	slices.SortFunc(a.Genres, func(a, b Genre) bool { return a.ID < b.ID })
 	a.Genres = slices.Compact(a.Genres)
+	a.Publisher = slice.MostFrequent(a.Publishers).Name
```
Comment (certuna) [inline model/mediafile.go:165]: Maybe I'm nitpicking, but wouldn't we want `a.Publisher` to be the one from the oldest album (i.e. earliest release date), not the most frequent?

File: ui/src/common/SongInfo.js [hunk 2]
```diff
@@ -49,10 +49,17 @@ export const SongInfo = (props) => {
     updatedAt: <DateField source="updatedAt" showTime />,
     playCount: <TextField source="playCount" />,
     bpm: <NumberField source="bpm" />,
+    publisher: <TextField source="publisher" />,
```
Comment (certuna) [inline ui/src/common/SongInfo.js:52]: This makes the (already space-constrained) songlist even more cramped, wouldn't it be better from a UI clutter point of view to have this column only available in the (sparser) albums list and leave it out for songs?

**Comment by certuna** [comment]
General comment: in most applications with support for the `TPUB`/`PUBLISHER`/`LABEL` field, they use the colloquial name "Record Label" (Plex and Roon for example). Do we want to use that in the UI, or stick to the more formal "Publisher"?

File: scanner/metadata/metadata.go [hunk 3]
```diff
@@ -103,6 +103,7 @@ func (t Tags) Year() int               { return t.getYear("date") }
 func (t Tags) Comment() string         { return t.getFirstTagValue("comment") }
 func (t Tags) Lyrics() string          { return t.getFirstTagValue("lyrics", "lyrics-eng") }
 func (t Tags) Compilation() bool       { return t.getBool("tcmp", "compilation") }
+func (t Tags) Publishers() []string    { return t.getAllTagValues("tpub", "publisher") }
```
[the remark this answers is not in the document]
Comment (raminou) [reply scanner/metadata/metadata.go]: fixed

File: model/mediafile.go (hunk 1 above)
Comment (raminou) [reply model/mediafile.go]: fixed

File: ui/src/common/SongInfo.js (hunk 2 above)
Comment (deluan) [reply ui/src/common/SongInfo.js:52]: These are not fields in the SongList, it is the information that is shown when you select "Get Info" from the context menu. I think this is the proper place to display this info

Comment (raminou) [reply ui/src/common/SongInfo.js:52]: I agree with you that the songlist is already very packed but the UI makes it very easy to hide. And I depends on everyone personal usage. In my case, I don't care about updatedAt or Channels fields, so I just hide those.

But if you prefer that, I am not completely against

Comment (deluan) [reply ui/src/common/SongInfo.js:52]: Maybe I was not clear in my previous comment: these fields are NOT SongList columns, they are shown here:
Screenshot 2023-05-18 at 1 21 50 PM

So this change is completely fine. And in the SongList you added it off by default so it is also ok.

File: scanner/metadata/metadata.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (raminou) [reply scanner/metadata/metadata.go]: I just saw that with "tbub" and "label", taglib returns 1 tag entry for each with the same value.

I will look at what taglib extracts and what can be done to workaround that.

[the remark this answers is not in the document]
Comment (deluan) [reply scanner/metadata/metadata.go]: I think it is safe to make `getAllTagValues` remove duplicated values.

[the remark this answers is not in the document]
Comment (raminou) [reply scanner/metadata/metadata.go]: Done

File: scanner/metadata/metadata_internal_test.go [hunk 4]
```diff
@@ -64,14 +64,22 @@ var _ = Describe("Tags", func() {
...
-		It("returns values from all tag names", func() {
+		It("returns genre values from all tag names", func() {
 			md := &Tags{}
 			md.tags = map[string][]string{
 				"genre": {"Rock", "Pop", "New Wave"},
 			}
 
 			Expect(md.Genres()).To(ConsistOf("Rock", "Pop", "New Wave"))
 		})
+		It("returns publisher values from all tag names", func() {
```
Comment (deluan) [inline scanner/metadata/metadata_internal_test.go:75]: This test is a duplicate of the previous one, and is not testing the dedup logic you implemented. Also, I think we should rename these two `It`s. Maybe something like:

```go
    It("returns values from all tag names")
    
    and
    
    It("remove duplicated values")
```

File: ui/src/i18n/en.json [hunk 5]
```diff
@@ -26,6 +27,7 @@
                 "rating": "Rating",
                 "quality": "Quality",
                 "bpm": "BPM",
+                "publisher": "Publisher",
```
Comment (deluan) [inline ui/src/i18n/en.json:30]: Duplicated

**Review by deluan — CHANGES_REQUESTED** [review]
Hey, thanks for this PR. I see you copied most of the implementation from Genre. That is fine, but we should avoid duplicate code and try to generalize what was done for genre and reuse in publisher. Examples:
- cached_publisher_repository.go
- sql_publisher.go
- mapPublishers in mapping.go

Also please remove any changes under `/subsonic` folder, as this is exclusive to Subsonic API implementation and it does not support publishers. 

Please also see my comments below. Thanks.

File: ui/src/album/AlbumDetails.js [hunk 6]
```diff
@@ -286,6 +287,9 @@ const AlbumDetails = (props) => {
             <Typography component={'div'} className={classes.recordMeta}>
               <Details />
             </Typography>
+            <Typography component={'p'} className={classes.recordMeta}>
```
[the remark this answers is not in the document]
Comment (certuna) [reply ui/src/album/AlbumDetails.js:290]: With the Original Date/Release Date PR now merged, it's probably better to create a new line with the release info we have: containing something like:

**♫ 1982-1986 · Release Date 10 January 2016 · Atlantic · ATLC-12345**

File: model/mediafile.go [hunk 7]
```diff
@@ -133,12 +136,16 @@ func (mfs MediaFiles) ToAlbum() Album {
 		if a.MinYear == 0 {
 			a.MinYear = m.Year
 		} else if m.Year > 0 {
-			a.MinYear = number.Min(a.MinYear, m.Year)
+			if m.Year < a.MinYear {
+				a.MinYear = m.Year
+				oldestPublisher = m.Publisher
```
[the remark this answers is not in the document]
Comment (raminou) [reply model/mediafile.go]: Fixed + Tests https://github.com/raminou/navidrome/blob/add-publisher/389/model/mediafile_test.go#L144

File: model/album.go [hunk 8]
```diff
@@ -75,10 +77,13 @@ func (als Albums) ToAlbumArtist() Artist {
 		a.SongCount += al.SongCount
 		a.Size += al.Size
 		a.Genres = append(a.Genres, al.Genres...)
+		a.Publishers = append(a.Publishers, al.Publishers...)
 		mbzArtistIds = append(mbzArtistIds, al.MbzAlbumArtistID)
 	}
 	slices.SortFunc(a.Genres, func(a, b Genre) bool { return a.ID < b.ID })
 	a.Genres = slices.Compact(a.Genres)
+	slices.SortFunc(a.Publishers, func(a, b Publisher) bool { return a.ID < b.ID })
+	a.Publishers = slices.Compact(a.Publishers)
```
[the remark this answers is not in the document]
Comment (raminou) [reply model/album.go:86]: Tests https://github.com/raminou/navidrome/blob/add-publisher/389/model/album_test.go#L67

File: persistence/album_repository.go [hunk 9]
```diff
@@ -140,6 +161,16 @@ func (r *albumRepository) GetAllWithoutGenres(options ...model.QueryOptions) (mo
 	return res, err
 }
 
+func (r *albumRepository) GetAllWithoutPublishers(options ...model.QueryOptions) (model.Albums, error) {
```
[the remark this answers is not in the document]
Comment (raminou) [reply persistence/album_repository.go]: Right I just removed it

File: scanner/metadata/metadata_internal_test.go (hunk 4 above)
Comment (raminou) [reply scanner/metadata/metadata_internal_test.go]: Done

File: ui/src/album/AlbumDetails.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (raminou) [reply ui/src/album/AlbumDetails.js]: I removed the duplicated display of the publisher.

I did not rebase yet, so I don't have all the Original/Release Date changes. I would like to have the other changes validated before squashing and rebasing

File: ui/src/i18n/en.json (hunk 5 above)
Comment (raminou) [reply ui/src/i18n/en.json]: Removed

File: ui/src/album/AlbumDetails.js [hunk 10]
```diff
@@ -206,6 +206,7 @@ const Details = (props) => {
         })}
     </>
   )
+  !isXsmall && addDetail(<>{record.publisher}</>)
```
[the remark this answers is not in the document]
Comment (raminou) [reply ui/src/album/AlbumDetails.js]: Modified

File: scanner/metadata/metadata.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (certuna) [reply scanner/metadata/metadata.go]: Actually, after testing, this is enough:

`func (t Tags) Publishers() []string    { return t.getAllTagValues("publisher", "label") }`

Taglib and ffmpeg both have an internal mapping for the `TPUB` id3 frame, so you don't have to query for the raw frame code.

**Comment by raminou** [comment]
> Hey, thanks for this PR. I see you copied most of the implementation from Genre. That is fine, but we should avoid duplicate code and try to generalize what was done for genre and reuse in publisher. Examples:
> 
>     * cached_publisher_repository.go
>     * sql_publisher.go
>     * mapPublishers in mapping.go

I understand that duplicated code is not the best way to do things. But do you think that other fields would be concern by this kind of changes ? Because in my opinion, it is not a big deal to have it duplicated, only 2 fields are in this case as far as I know.

In the case you still would like to have it, do you have any idea of how to name a group with genres and publishers (and maybe others) ?

**Comment by certuna** [comment]
> I understand that duplicated code is not the best way to do things. But do you think that other fields would be concern by this kind of changes ? Because in my opinion, it is not a big deal to have it duplicated, only 2 fields are in this case as far as I know.

Generic functions would be very useful (necessary even) once we add other many-to-many relations like Language, Release Type and Mood tags, and even multiple artists.

**Comment by deluan** [comment]
> Generalized code would be very useful (necessary even) to add other many-to-many relations like Language, Release Type and Mood tags, and even multiple artists.
> 
> Maybe use `mapM2m`, `sql_m2m.go`, `cached_m2m_repository.go`, what do you think @deluan ?

Yes. Another example that would be exactly the same is `grouping`. But I think the current solution for M2M tags is not optimal, as they increase execution time for scanning and querying the DB. I may refactor this to maybe use JSON objects in the DB.

@raminou I think it is ok to keep the duplication in this PR, I'll work on that later.

File: ui/src/album/AlbumDetails.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (deluan) [reply ui/src/album/AlbumDetails.js]: I think it is in a good shape now. Can you please rebase it?

File: scanner/metadata/metadata.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (raminou) [reply scanner/metadata/metadata.go]: I would say that it does not cost anything to keep it, to not depend of the internal mapping of any tag extractor

File: ui/src/album/AlbumDetails.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (raminou) [reply ui/src/album/AlbumDetails.js]: I have 
image

The files are not using release date/original date, only the classic "Year".

[the remark this answers is not in the document]
Comment (certuna) [reply ui/src/album/AlbumDetails.js]: I'm taking the case for albums where Original Date, Date/Year and Release Date are all different, for example most classical music albums

[the remark this answers is not in the document]
Comment (certuna) [reply ui/src/album/AlbumDetails.js]: Screenshot 2023-05-30 at 16 42 18

**Comment by deluan** [comment]
> General comment: in most applications with support for the `TPUB`/`PUBLISHER`/`LABEL` field, they use the colloquial name "Record Label" (Plex, Kodi and Roon for example). Do we want to use that in the UI, or stick to the more formal "Publisher"?

I think I prefer "Record Label" (UX's principle of least effort). What do you think @raminou ?

@certuna, any other comments on the PR as a whole? Besides the naming above, are we code complete?

**Comment by raminou** [comment]
I replaced the name "Publisher" by "Record Label" as requested.

And rebased from master

**Comment by deluan** [comment]
Hey @raminou, thanks for your effort, but I decided to go in a different direction, and soon I'll enable multi-valued tags in a faster and simpler way. Thanks anyway.

**Pull request closed without merging**